File: api/src/utils/auth0.ts
/**
 * Discord API utilities.
 */

// External imports.
import axios from "axios"

// Utility imports.
import { redis } from "@utils/redis"

/**
 * Fetch a management API token.
 */
interface FetchManagementTokenOutput {
  accessToken: string
  expires: number
}
async function fetchManagementToken(): Promise<FetchManagementTokenOutput> {
  const response = await axios({
    method: "POST",
    baseURL: "https://discordbuilder.us.auth0.com",
    url: "/oauth/token",
    headers: {
      "Content-Type": "application/json",
    },
    data: {
      client_id: process.env.AUTH0_CLIENT_ID,
      client_secret: process.env.AUTH0_CLIENT_SECRET,
      audience: "https://discordbuilder.us.auth0.com/api/v2/",
      grant_type: "client_credentials",
    },
  })
  const accessToken = response.data.access_token
  const expires =
    JSON.parse(
      Buffer.from(accessToken.split(".")[1], "base64").toString("utf-8"),
    ).exp * 1000
  return { accessToken, expires }
}

/**
 * Get management token.
 * This wraps the fetchManagementToken function to cache received tokens.
 */
let _cachedTokenResponse: FetchManagementTokenOutput | undefined
async function getManagementToken(): Promise<string> {
  if (
    !_cachedTokenResponse ||
    _cachedTokenResponse.expires < Date.now() - 1 * 60 * 60 * 1000
  ) {
    _cachedTokenResponse = await fetchManagementToken()
  }
  return _cachedTokenResponse.accessToken
}

/**
 * Get Discord token from Auth0.
 */
interface GetDiscordTokenFromAuth0Input {
  sub: string
}
async function getDiscordTokenFromAuth0({
  sub,
}: GetDiscordTokenFromAuth0Input): Promise<DiscordToken | undefined> {
  try {
    const token = await getManagementToken()
    const response = await axios({
      method: "GET",
      baseURL: "https://discordbuilder.us.auth0.com/api/v2",
      url: `/users/${sub}`,
      headers: {
        Authorization: `Bearer ${token}`,
      },
    })
    if (response.status !== 200) return
    if (!response.data) return
    const identities: any[] | undefined = response.data.identities
    if (!identities?.length) return
    const identity: {
      access_token?: string
      refresh_token?: string
    } = identities.find((x) => x.connection === "discord")
    if (!identity) return
    if (!identity.access_token) return
    if (!identity.refresh_token) return
    return {
      accessToken: identity.access_token,
      refreshToken: identity.refresh_token,
    }
  } catch (error) {
    // TODO: Log error.
    return
  }
}

/**
 * Get Discord token from database.
 */
interface GetDiscordTokenFromDBInput {
  sub: string
}
async function getDiscordTokenFromDB({
  sub,
}: GetDiscordTokenFromDBInput): Promise<DiscordToken | undefined> {
  const accessToken = await redis.hget(`admin:idp:${sub}`, "accessToken")
  if (!accessToken) return
  const refreshToken = await redis.hget(`admin:idp:${sub}`, "refreshToken")
  if (!refreshToken) return
  return { accessToken, refreshToken }
}

/**
 * Set Discord token in database.
 */
interface SetDiscordTokenInDBInput {
  sub: string
  token: DiscordToken
}
async function setDiscordTokenInDB({ sub, token }: SetDiscordTokenInDBInput) {
  await redis.hset(`admin:idp:${sub}`, {
    accessToken: token.accessToken,
    refreshToken: token.refreshToken,
  })
}

/**
 * Fetch a users Discord token from Auth0.
 */
interface FetchDiscordTokenInput {
  sub: string
}
export interface DiscordToken {
  accessToken: string
  refreshToken: string
}
interface FetchDiscordTokenOutput {
  token?: DiscordToken
}
export async function fetchDiscordToken({
  sub,
}: FetchDiscordTokenInput): Promise<FetchDiscordTokenOutput> {
  const dbToken = await getDiscordTokenFromDB({ sub })
  if (!dbToken) {
    const idpToken = await getDiscordTokenFromAuth0({ sub })
    if (idpToken) {
      await setDiscordTokenInDB({ sub, token: idpToken })
    }
    return { token: idpToken }
  }
  return { token: dbToken }
}

File: bot/builder/utils/auth0.py
"""Auth0 utilities.
"""

# System imports.
import os
import time
import aiohttp
import json
import base64
from typing import Optional, Tuple

# Utility imports.
from builder.utils import Singleton


class __TokenSingleton(Singleton):
    """Singleton to handle storing tokens and expirations.
    """

    expires: int = 0
    token: str = ""

    async def get_token(self) -> str:
        """Wrapper to return cached token or fresh.
        """

        # Fetch a new token if the old one expired.
        if (not self.token or self.expires > int(time.time()) - 60):
            expires, token = await self.__fetch_token()
            self.expires = expires
            self.token = token

        # Return the token.
        return self.token

    async def __fetch_token(self) -> Tuple[int, str]:
        """Fetch an API token from Auth0.
        """

        # Check configuration.
        client_id = os.environ.get("AUTH0_CLIENT_ID")
        client_secret = os.environ.get("AUTH0_CLIENT_SECRET")
        if (not client_id or not client_secret):
            return 0, None

        # Fetch a new token.
        async with aiohttp.ClientSession() as session:
            async with session.post(
                "https://discordbuilder.us.auth0.com/oauth/token",
                json={
                    "client_id": client_id,
                    "client_secret": client_secret,
                    "audience": "builder",
                    "grant_type": "client_credentials",
                }
            ) as response:
                json_body = await response.json()
                token: Optional[str] = json_body.get("access_token")

        # Throw if no token.
        if not token:
            raise Exception("Failed to retrieve token.")

        # Determine expiration.
        payload = json.loads(base64.b64decode(token.split(".")[1]))
        expires: int = payload.get("exp", 0)

        # Return the responses.
        return expires, token


async def get_token():
    """Wrapper to get a token from Auth0.
    """

    # Return the token.
    auth0 = __TokenSingleton()
    return await auth0.get_token()
